Add a printer that knows how to print user-defined columns

brendandburns opened the pull request:
@bgrant0607 @jlowdermilk @JanetKuo @deads2k 

This was discussed in the context of https://github.com/kubernetes/kubernetes/issues/7843

I can't find a dedicated issue for it, but I swear we discussed it.

I'll integrate this as a supported printer in a follow-on PR.

brendandburns committed:
Add a printer that knows how to print user-defined columns

deads2k commented on pkg/kubectl/custom_column_printer.go:
```diff
@@ -0,0 +1,92 @@
+
+func (s *CustomColumnsPrinter) PrintObj(obj runtime.Object, out io.Writer) error {
+	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
+	buff := &bytes.Buffer{}
```
I don't see this used.

brendandburns replied:
oops, first implementation cruft, removed.

deads2k commented on pkg/kubectl/custom_column_printer.go:
```diff
@@ -0,0 +1,92 @@
+	for ix := range s.Columns {
+		fmt.Fprintf(w, "%s\t", s.Columns[ix].Header)
+	}
+	fmt.Fprintln(w, buff.String())
```
This is just `fmt.Fprintln(w)`, right?

brendandburns replied:
as above, fixed.

deads2k commented on pkg/kubectl/custom_column_printer.go:
```diff
@@ -0,0 +1,92 @@
+			return err
+		}
+		if len(values) != 1 || len(values[0]) != 1 {
+			return fmt.Errorf("unexpected number of values: %d %v", len(values), values)
```
I'm not certain that I always want this condition to fail hard.  If I want to list out the first, second, and third items in a list (say the secrets on a serviceaccount), not all serviceaccounts will have at least three, but I probably don't want my printer to fail to display anything to me.

brendandburns replied:
Much improved. thanks for the comment.

bgrant0607 commented:
See also #11695

cc @kubernetes/kubectl @mwielgus

mwielgus commented:
Can we combine HumanReadablePrinter (from #11695) and CustomColumnsPrinter into one?

brendandburns commented:
Yeah, I think that's the right thing to do in the limit, have a bunch of template files as @bgrant0607 suggests and refactor HumanReadablePrinter in terms of the CustomColumnsPrinter.

Why don't we merge in the custom columns printer, and then you can refactor #11695 in terms of that printer?  Does that work for you?

brendandburns commented:
Comments addressed, please re-check.

Thanks!
--brendan

deads2k commented:
lgtm.  Weird that `test-cmd` would only fail on go 1.3.  I don't have re-run rights, can you kick shippable?

bgrant0607 commented on pkg/kubectl/custom_column_printer_test.go:
```diff
@@ -0,0 +1,91 @@
+			columns: []Column{
+				{
+					Header:    "NAME",
+					FieldSpec: "{.ObjectMeta.Name}",
```
These expressions should use the json field names, not the Go field names.

bgrant0607 replied:
They also need to reference the versioned APIs.

deads2k replied:
> They also need to reference the versioned APIs.

That being the case, the `SortingPrinter` from https://github.com/kubernetes/kubernetes/pull/12432 needs updating as well.

brendandburns replied:
How do you imaging referencing versioned apis should work?  I don't think that's right.  I think that the templates themselves may be versioned, and the user will select, but in this context, I don't think that versioning has a place.

bgrant0607 replied:
Go field names are an implementation detail that we shouldn't expose to users. The json field names are the API, and all such API resources and their fields are versioned. We already do this with the Go templates.

bgrant0607 replied:
Good call, @deads2k. Yes, the sorting feature does need to be fixed.

bgrant0607 commented:
This implementation uses the jsonpath template syntax, which is more usable for raw fields than the Go template syntax, but is insufficient for the HumanReadablePrinter use case.

Note: jsonpath template support is being added to kubectl in PR #12202.

bgrant0607 commented:
Any thoughts about how this would be exposed through kubectl? Assuming we continue to support both Go templates and jsonpath templates, we'll want this functionality for both template schemes.

brendandburns commented:
The jsonPath should use json names, but that requires changes to the JSONPath libraries reflection code.

Why is this insufficient for human readable printer?  Just for human readable numbers like age, etc.

brendandburns commented:
For kubectl, I think that we want to both have:
`kubectl ... --custom-columns=<template>:<header>,<template>:<header>,<template>:<header>`

as well as:

`kubectl ... --custom-columns-template=<template-file>`

bgrant0607 commented:
TestPrintJSONPath in #12202 prints `"{.metadata.name}"`, which refers to the json fields of a versioned API resource.

bgrant0607 commented:
Both from command line and file make sense, but should be output format options, not new flags.

bgrant0607 commented:
As for HumanReadablePrinter, many of the fields (e.g., time fields) are transformed for output. How would such transformations be specified when using jsonpath templates?

brendandburns commented:
For jsonpath templates, we could either have users specify explicit output transformations (kind of ackward) or infer them from the type of the value (far easier to use)

Either way, it's the right approach from technical organization.

Ah, I think I misinterpretted your comments, I think you meant: "You should run templates against versioned objects"  not "the templates should include the API version"

I think that's probably true, but it is a different from other printers, we'll have to think about the right way to expose this.

brendandburns commented:
Ok, this is fixed to use json names, and I fixed the sorting printer too (and I fixed up some hacky code in the jsonPath printer that I should have caught before...)

Please take another look.

Thanks!
--brendan

bgrant0607 commented on pkg/kubectl/custom_column_printer_test.go:
```diff
@@ -0,0 +1,91 @@
+			columns: []Column{
+				{
+					Header:    "NAME",
+					FieldSpec: "{.ObjectMeta.Name}",
```
Why are these not the json field names?

brendandburns replied:
oops, oversight.  fixed.

bgrant0607 commented:
It shouldn't be different than -o template. That definitely uses the versioned json field names. We had to add --api-version= to all our scripts that used templates to prevent them from breaking repeatedly when changing the default API version.

brendandburns commented:
@bgrant0607 yeah, I guess you're right, I always thought we took it into the in-memory representation and then printed them out, but after looking at the code, I realize I'm wrong.

Anyway, this works correctly now (as does the sorting printer)

and `jsonpath` should now correctly handle `inline` fields, which I'm not sure it did before...

bgrant0607 commented:
@jlowdermilk PTAL

jlowdermilk commented on pkg/kubectl/custom_column_printer.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func (s *CustomColumnsPrinter) PrintObj(obj runtime.Object, out io.Writer) error {
+	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
```
Suggest making these params constants so it's easy to see what each one is.

kargakis replied:
+1

brendandburns replied:
done.

jlowdermilk commented on pkg/util/jsonpath/jsonpath.go:
```diff
@@ -258,6 +259,38 @@ func (j *JSONPath) evalUnion(input []reflect.Value, node *UnionNode) ([]reflect.
+		}
+	}
+	if inlineValue != nil {
+		fmt.Printf("inline value while looking for: %s\n", node.Value)
```
This looks like a debug line.

brendandburns replied:
removed.

brendandburns commented:
Comments addressed, @jlowdermilk ptal

thanks
--brendan

brendandburns commented:
nm, I'm fixing the test...

jlowdermilk commented:
lgtm, please squash

brendandburns commented:
squashed.

brendandburns commented:
LGTM was before last commit, removing LGTM

brendandburns commented:
last push was a rebase...  Re-adding LGTM

brendandburns commented:
@k8s-bot test this please

brendandburns commented:
@k8s-bot test this please

jlowdermilk commented:
lgtm

kargakis commented on pkg/kubectl/custom_column_printer.go:
```diff
@@ -0,0 +1,107 @@
+			return err
+		}
+	}
+	return w.Flush()
```
Shouldn't this flush be deferred?

brendandburns commented:
merging, since submit-queue skipped this PR due to github API issues.
